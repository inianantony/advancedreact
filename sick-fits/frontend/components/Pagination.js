import React, { Component } from 'react';
import PaginationStyles from './styles/PaginationStyles';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import { perPage } from '../config'
import Head from 'next/head'
import Link from 'next/link'

const PAGINATION_QUERY = gql`
    query PAGINATION_QUERY {
        itemsConnection {
            aggregate{
                count
            }
        }
    }
`;

const Pagination = props => (

    <Query query={PAGINATION_QUERY}>
        {({ error, loading, data }) => {
            if (error) return <p>Error...</p>
            if (loading) return <p>Loading...</p>
            const count = data.itemsConnection.aggregate.count;
            const pages = Math.ceil(count / perPage);
            return (
                <PaginationStyles>
                    <Head>
                        <title>Sick Fits | Page {props.page} of {pages} </title>
                    </Head>
                    <Link prefetch href={{ path: './items', query: { page: props.page - 1 } }}>
                        <a aria-disabled={props.page<=1}>⬅ Prev</a>
                    </Link>
                    <p>Page {props.page} of {pages}</p>
                    <p>{count} Items Total</p>
                    <Link prefetch href={{ path: './items', query: { page: props.page + 1 } }}>
                        <a aria-disabled={props.page>=pages}>Next ➡</a>
                    </Link>
                </PaginationStyles>
            )
        }}
    </Query>


);

export default Pagination;