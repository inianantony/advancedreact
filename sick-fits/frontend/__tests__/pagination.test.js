import { mount } from 'enzyme';
import wait from 'waait';
import Router from 'next/router';
import Pagination, { PAGINATION_QUERY } from '../components/Pagination';
import { MockedProvider } from 'react-apollo/test-utils';
import toJSON from 'enzyme-to-json';
Router.router = {
    push(){},
    prefetch(){}
}

function makeMocksFor(length) {
    return [{
        request: { query: PAGINATION_QUERY },
        result: { data: { itemsConnection:{
            __typename: 'aggregate',
            aggregate:{
                __typename: 'count',
                count: length
            }
        } } }
    }];
}

describe("<Pagination/>", () => {
    it("displays a loading message", () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(1)}>
                <Pagination page={1}/>
            </MockedProvider>
        );
        expect(wrapper.text()).toContain("Loading...");
    });

    it("renders pagination for 18 items", async () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(18)}>
                <Pagination page={1}/>
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        expect(toJSON(wrapper.find("p"))).toMatchSnapshot();
        expect(toJSON(wrapper.find("Link"))).toMatchSnapshot();
    });

    it("disables prev button on first page", async () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(18)}>
                <Pagination page={1}/>
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        expect(wrapper.find("a.prev").prop('aria-disabled')).toBe(true);
        expect(wrapper.find("a.next").prop('aria-disabled')).toBe(false);
    });

    it("disables next button on last page", async () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(18)}>
                <Pagination page={5}/>
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        expect(wrapper.find("a.prev").prop('aria-disabled')).toBe(false);
        expect(wrapper.find("a.next").prop('aria-disabled')).toBe(true);
    });

    it("enables all buttons on middle page", async () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(18)}>
                <Pagination page={3}/>
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        expect(wrapper.find("a.prev").prop('aria-disabled')).toBe(false);
        expect(wrapper.find("a.next").prop('aria-disabled')).toBe(false);
    });
});