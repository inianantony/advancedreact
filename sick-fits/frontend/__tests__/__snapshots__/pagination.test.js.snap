// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination/> renders pagination for 18 items 1`] = `
Array [
  <p>
    Page 
    1
     of 
    5
  </p>,
  <p>
    18
     Items Total
  </p>,
]
`;

exports[`<Pagination/> renders pagination for 18 items 2`] = `
Array [
  <Link
    href={
      Object {
        "path": "./items",
        "query": Object {
          "page": 0,
        },
      }
    }
    prefetch={true}
  >
    <a
      aria-disabled={true}
      className="prev"
      href="?page=0"
      onClick={[Function]}
    >
      ⬅ Prev
    </a>
  </Link>,
  <Link
    href={
      Object {
        "path": "./items",
        "query": Object {
          "page": 2,
        },
      }
    }
    prefetch={true}
  >
    <a
      aria-disabled={false}
      className="next"
      href="?page=2"
      onClick={[Function]}
    >
      Next ➡
    </a>
  </Link>,
]
`;
